src/NumberTheory/vinberg.jl

```diff
@@ -0,0 +1,295 @@
+  for v in possible_vec
+    vQ = v * Q
+    n = (vQ*v1)[1, 1]
+    k = (vQ*transpose(v))[1, 1]
+    push!(possible_vec_sort, (v, n, k))
+  end
+  sort!(possible_vec_sort; by=x -> (x[2]^2) // abs(x[3]))
+
+  # We execute the algorithm with replacing v0 by v1
+  for (v, n) in possible_vec_sort
+    if n < 0
+      v = -v
+    end
+    if all((v*Q*transpose(w))[1, 1] >= 0 for w in roots)
+
+      push!(roots, v)
+    end
+  end
+  return roots
+end
+
+@doc raw"""
+    _generate_direction_vector(Q::ZZMatrix, v0::ZZMatrix, possible_vec::Vector{ZZMatrix}) -> ZZMatrix
+
+Since no direction vector v1 was given, we need to find one which satisfies the following conditions:
+- v0.v1 = 0
+- v.v1 != 0 for all possible roots v with v.v0 = 0
+"""
+function _generate_direction_vector(Q::ZZMatrix, v0::ZZMatrix, possible_vec::Vector{ZZMatrix})
+  l = length(v0)
+  v1 = zero_matrix(QQ, l, 1) # initializing v1 that it survives the for loop
+  signal = 1 # initializing a stopping condition
+  while signal in 1:100000000
+    if l < 4
+      v1_ = rand(-30:30, l)
+    elseif l < 10
+      v1_ = rand(-25:25, l)
+    else
+      v1_ = rand(-20:20, l)  # for higher l it is not necessary/efficient to choose a big random range
+    end
+    v1 = matrix(QQ, l, 1, v1_)
+    if (v0*Q*v1)[1, 1] != 0
+      signal += 1
+      continue
+    end
+    @hassert :Vinberg 1 (transpose(v1)*Q*v1)[1, 1] < 0
+    if all((v*Q*v1)[1, 1] != 0 for v in possible_vec)
+      # To run the algorithm it suffices to require that v*Q*v1 != 0, but we get a random (right) solution. 
+      # This is the case because it is a random choice in which direction the fundamental cone 
+      # is built everytime we use the algorithm.
+      # To get always the same solution one could require that v*Q*v1 > 0, but this has very huge impact on the efficiency. 
+      @vprintln :Vinberg 2 "direction vector v1 = $v1"
+      return v1
+    end
+    signal += 1
+  end
+  @req signal < 10000000 "Choose another v0" # break the algorithm if no v1 was found
+end
+
+@doc raw"""
+    _check_direction_vector(Q::ZZMatrix, v0::ZZMatrix, possible_vec::Vector{ZZMatrix}, direction_vector::ZZMatrix) -> ZZMatrix
+
+We check if the given direction vector holds the following requirements:
+- v0.v1 = 0
+- v.v1 != 0 for all possible roots v with v.v0 = 0
+"""
+function _check_direction_vector(Q::ZZMatrix, v0::ZZMatrix, possible_vec::Vector{ZZMatrix}, direction_vector::ZZMatrix)
+  v1 = transpose(direction_vector)
+  @req (v0*Q*v1)[1, 1] == 0 "The direction vector has to be orthogonal on v0"
+  @req all((v*Q*v1)[1, 1] != 0 for v in possible_vec) "The direction vector cannot be orthogonal on one of the possible roots"
+  return v1
+end
+
+@doc raw"""
+    vinberg_algorithm(Q::ZZMatrix, upper_bound::ZZRingElem; v0::ZZMatrix, root_lengths::Vector{ZZRingElem}, direction_vector::ZZMatrix) -> Vector{ZZMatrix}
+
+Return the roots r of a given hyperbolic reflection lattice Q with squared length contained in root_lengths and 
```

Comment:
The keywords such as `r` should be marked as code, i.e.:
```suggestion
Return the roots `r` of a given hyperbolic reflection lattice `Q` with squared length contained in `root_lengths` and 
```
and so on in the lins below.